Dev/qp

cheesecakeball opened the pull request:
Add DspOsiGrb.h

kibaekkim commented:
@cheesecakeball need `try` and `catch` to handle `GUROBI_CALL`.

kibaekkim commented on src/SolverInterface/DspOsiGrb.h:
```diff
@@ -0,0 +1,168 @@
+        case GRB_TIME_LIMIT:
+            status = DSP_STAT_STOPPED_TIME;
+            break;
+        case GRB_NUMERIC:
```
This should be `DSP_STAT_ABORT`.

kibaekkim commented:
@cheesecakeball Please test the code first.

cheesecakeball commented:
ready to review

kibaekkim commented on CMakeLists.txt:
```diff
@@ -51,6 +51,7 @@ endif()
 find_library(SCIPLIB scip PATHS ${SCIPOPT_LIB_DIR})
 find_library(MA27LIB  ma27  PATHS ${MA27LIB_DIR})
 find_library(CPLEXLIB cplex PATHS ${CPLEX_LIB_DIR})
+find_library(GUROBILIB gurobi${GUROBI_VERSION} PATHS ${GUROBI_LIB_DIR})
```
Where do we get `GUROBI_VERSION`?

cheesecakeball replied:
Gurobi library is defined as `libgurobi90.dylib` on mac. "90" means version 9.0. It is different from CPLEX library `libcplex.a`.

cheesecakeball replied:
The code has been tested. I will resolve the confilict.

kibaekkim replied:
> The code has been tested. I will resolve the confilict.

Please see this: https://github.com/Argonne-National-Laboratory/DSP/pull/63#issuecomment-634934107

kibaekkim replied:
> Gurobi library is defined as `libgurobi90.dylib` on mac. "90" means version 9.0. It is different from CPLEX library `libcplex.a`.

**WHERE** does variable `GUROBI_VERSION` set?

cheesecakeball replied:
`GUROBI_VERSION` is set in `UserConfig.cmake`. Just pushed this file, too.

kibaekkim replied:
Please consider to use this: https://support.gurobi.com/hc/en-us/articles/360039499751-CMake-C-C-compilation-of-Gurobi-projects

cheesecakeball replied:
> Please consider to use this: https://support.gurobi.com/hc/en-us/articles/360039499751-CMake-C-C-compilation-of-Gurobi-projects

This is the file I refer to. In Gurobi, the library files are defined [here](https://www.gurobi.com/documentation/9.0/quickstart_mac/file_overview.html). `libgurobi90.so` and `libgurobi90.dylib` are libraries for C interface on Windows and Mac, respectively.

In the `FindGUROBI.cmake` file in the quoted link, they have the following command to find the C library

```
find_library(GUROBI_LIBRARY
    NAMES gurobi gurobi81 gurobi90
    HINTS ${GUROBI_DIR} $ENV{GUROBI_HOME}
    PATH_SUFFIXES lib)
```
which is the same as what I did.

kibaekkim replied:
Would not it be better to list names `gurobi` `gurobi81` `gurobi90` as in their example?

cheesecakeball replied:
I can also do that, but we may need to update this line in CMake files every time when Gurobi has new versions. I think the way I did is more general.

kibaekkim replied:
I prefer updating cmake file over asking users to put extra information. In the sense, gurobi example is more user friendly.

cheesecakeball replied:
Sure. Then I will update the CMAKE file.

kibaekkim commented on src/CMakeLists.txt:
```diff
@@ -88,6 +86,19 @@ if(CPLEXLIB)
+    set(DSP_LIBS
+        ${DSP_LIBS}
+        ${DEPEND_DIR}/lib/libOsiGrb.a
+        ${GUROBI_LIB_DIR}/libgurobi90.dylib)
```
The library name should not be hard coded.

kibaekkim commented on src/CMakeLists.txt:
```diff
@@ -168,9 +179,12 @@ if (MPI_CXX_FOUND)
 set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -Wno-absolute-value")
 
+=======
+>>>>>>> upstream/master
```
remove garbage lines

cheesecakeball commented:
The code is ready to review. The test was passed.

kibaekkim commented on src/CMakeLists.txt:
```diff
@@ -86,6 +86,19 @@ if(CPLEXLIB)
+    set(DSP_LIBS
+        ${DSP_LIBS}
+        ${DEPEND_DIR}/lib/libOsiGrb.a
+        ${GUROBI_LIB_DIR}/libgurobi${GUROBI_VERSION}.dylib)
```
`.dylib` is specific to macosx. Please make it generic.

kibaekkim commented on src/Utility/DspParams.cpp:
```diff
@@ -212,25 +210,27 @@ void DspParams::initIntParams()
 #endif
 
+#ifdef DSP_HAS_GRB
+	IntParams_.createParam("DE/SOLVER", OsiGrb);
```
You should have received a warning message if you have gruobi and other solver, because the same parameter is defined twice.

kibaekkim commented on UserConfig.cmake:
```diff
@@ -1,7 +1,10 @@
 set(MA27LIB_DIR     "")
-set(CPLEX_LIB_DIR   "")
-set(CPLEX_INC_DIR   "")
+set(CPLEX_LIB_DIR	"/Applications/CPLEX_Studio129/opl/lib/x86-64_osx/static_pic")
```
These variables should be set to "" in the file.

kibaekkim commented on extra/CMakeLists.txt:
```diff
@@ -74,6 +72,15 @@ else()
 endif()
 
+if(GUROBILIB)
+	set(GUROBI_LIB "-L${GUROBI_LIB_DIR} -lgurobi${GUROBI_VERSION} -lm -lpthread")
```
The library name should be captured by cmake variable. Try `GUROBILIB` or find the relevant cmake manual.

cheesecakeball commented:
This is ready to review, but I have some issues here.

In the [param_grb.txt](https://github.com/cheesecakeball/DSP/blob/fa2b5ec02ab53a2de5658c600478959ac7da04c0/test/params_grb.txt#L8), if I set the parameter `DW/MASTER/SOLVER` to 0 as in the file, I can pass the test. However, if I am using Gurobi to solve DW master problem, i.e., `DW/MASTER/SOLVER 4`, I could not pass the test. I got the following result

```
Creating DSP environment
Reading SMPS files: farmer
Coin0001I At line 2 NAME          FARMER
Coin0001I At line 3 ROWS
Coin0001I At line 9 COLUMNS
Coin0001I At line 22 RHS
Coin0001I At line 25 BOUNDS
Coin0001I At line 30 ENDATA
Coin0002I Problem FARMER has 4 rows, 9 columns and 12 elements
Coin0001I At line 2 TIME          FARMER
Coin0001I At line 3 PERIODS       IMPLICIT
Coin0001I At line 6 ENDATA

Coin0001I At line 2 STOCH         FARMER
Coin0001I At line 3 SCENARIOS
Coin0001I At line 16 ENDATA
Read SMPS files: 0.000808 sec.
First stage: 1 rows, 3 cols, 3 integers
Second stage: 3 rows, 6 cols, 0 integers
Number of scenarios: 3
Reading parameter files: /Users/yingqiuzhang/Documents/DSPtest/DSP/test/params_grb.txt
Initializing subproblems ... 
Academic license - for non-commercial use only
Warning: parameter changes on this environment will not affect existing models.
Initializing master problem ... 
Initializing ALPS framework ... 
DwModelSmip constructor.
Created BRANCH_NONANT2 rule.
==  Welcome to the Abstract Library for Parallel Search (ALPS) 
==  Copyright 2000-2017 Lehigh University and others 
==  All Rights Reserved. 
==  Distributed under the Eclipse Public License 1.0 
==  Version: Trunk (unstable) 
==  Build Date: Jun  2 2020
Alps0250I Starting search ...
Generated 3 initial columns. Initial dual bound -1.153999994455e+05
Iteration   0: DW Bound +1.797693e+308, Best Dual -1.154000e+05 (gap Large %), nrows 6, ncols 12, timing (total 0.01, master 0.00, gencols 0.01), statue 3000
Iteration   1: DW Bound +1.797693e+308, Best Dual -1.129653e+05 (gap Large %), nrows 9, ncols 12, timing (total 0.01, master 0.00, gencols 0.01), statue 3000
Iteration   2: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 11, ncols 12, timing (total 0.02, master 0.00, gencols 0.01), statue 3000
Iteration   3: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.02, master 0.00, gencols 0.02), statue 3000
Iteration   4: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.02, master 0.00, gencols 0.02), statue 3000
Iteration   5: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.03, master 0.00, gencols 0.02), statue 3000
Iteration   6: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.03, master 0.00, gencols 0.03), statue 3000
Iteration   7: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.03, master 0.00, gencols 0.03), statue 3000
Iteration   8: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.04, master 0.00, gencols 0.03), statue 3000
Iteration   9: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.04, master 0.00, gencols 0.04), statue 3000
Iteration  10: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.05, master 0.01, gencols 0.04), statue 3000
Iteration  11: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.05, master 0.01, gencols 0.04), statue 3000
Iteration  12: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.06, master 0.01, gencols 0.05), statue 3000
Iteration  13: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.06, master 0.01, gencols 0.05), statue 3000
Iteration  14: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.06, master 0.01, gencols 0.05), statue 3000
Iteration  15: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.07, master 0.01, gencols 0.06), statue 3000
Iteration  16: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.08, master 0.01, gencols 0.07), statue 3000
Iteration  17: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.08, master 0.01, gencols 0.07), statue 3000
Iteration  18: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.08, master 0.01, gencols 0.07), statue 3000
Iteration  19: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.08, master 0.01, gencols 0.07), statue 3000
Iteration  20: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.09, master 0.01, gencols 0.07), statue 3000
Iteration  21: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.09, master 0.01, gencols 0.08), statue 3000
Iteration  22: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.09, master 0.01, gencols 0.08), statue 3000
Iteration  23: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.10, master 0.01, gencols 0.08), statue 3000
Iteration  24: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.10, master 0.01, gencols 0.09), statue 3000
Iteration  25: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.11, master 0.01, gencols 0.09), statue 3000
Iteration  26: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.11, master 0.01, gencols 0.10), statue 3000
Iteration  27: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.12, master 0.01, gencols 0.10), statue 3000
Iteration  28: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.12, master 0.01, gencols 0.11), statue 3000
Iteration  29: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.13, master 0.01, gencols 0.11), statue 3000
Iteration  30: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.13, master 0.01, gencols 0.11), statue 3000
Iteration  31: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.14, master 0.01, gencols 0.12), statue 3000
Iteration  32: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.14, master 0.01, gencols 0.12), statue 3000
Iteration  33: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.14, master 0.01, gencols 0.13), statue 3000
Iteration  34: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.15, master 0.01, gencols 0.13), statue 3000
Iteration  35: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.15, master 0.01, gencols 0.13), statue 3000
Iteration  36: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.15, master 0.01, gencols 0.13), statue 3000
Iteration  37: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.15, master 0.01, gencols 0.14), statue 3000
Iteration  38: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.16, master 0.01, gencols 0.14), statue 3000
Iteration  39: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.16, master 0.01, gencols 0.14), statue 3000
Iteration  40: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.17, master 0.01, gencols 0.15), statue 3000
Iteration  41: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.17, master 0.01, gencols 0.15), statue 3000
Iteration  42: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.17, master 0.01, gencols 0.15), statue 3000
Iteration  43: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.17, master 0.02, gencols 0.16), statue 3000
Iteration  44: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.18, master 0.02, gencols 0.16), statue 3000
Iteration  45: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.18, master 0.02, gencols 0.16), statue 3000
Iteration  46: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.19, master 0.02, gencols 0.17), statue 3000
Iteration  47: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.19, master 0.02, gencols 0.17), statue 3000
Iteration  48: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.20, master 0.02, gencols 0.18), statue 3000
Iteration  49: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.20, master 0.02, gencols 0.18), statue 3000
Iteration  50: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.20, master 0.02, gencols 0.18), statue 3000
Iteration  51: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.21, master 0.02, gencols 0.18), statue 3000
Iteration  52: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.21, master 0.02, gencols 0.19), statue 3000
Iteration  53: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.21, master 0.02, gencols 0.19), statue 3000
Iteration  54: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.22, master 0.02, gencols 0.20), statue 3000
Iteration  55: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.22, master 0.02, gencols 0.20), statue 3000
Iteration  56: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.22, master 0.02, gencols 0.20), statue 3000
Iteration  57: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.23, master 0.02, gencols 0.20), statue 3000
Iteration  58: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.23, master 0.02, gencols 0.21), statue 3000
Iteration  59: DW Bound +1.797693e+308, Best Dual -1.120503e+05 (gap Large %), nrows 12, ncols 12, timing (total 0.24, master 0.02, gencols 0.21), statue 3000
Found new primal solution: -1.072400e+05
[0.240335] curLb 1.79769313e+308, bestUb -1.07239999e+05, bestLb 1.79769313e+308
The current node is fathomed.
Alps0240I Proc: 1, Part: 0, Cand: 0, Best N: 1e+75, Best S: -107240

Alps0208I Search completed.
Alps0260I Best solution found had quality -107240
Alps0264I Number of nodes processed:                1
Alps0267I Number of nodes branched:                 0
Alps0268I Number of nodes pruned before processing: 0
Alps0270I Number of nodes left:                     0
Alps0272I Tree depth: 0
Alps0274I Search CPU time: 0.19 seconds
Alps0278I Search wall-clock time: 0.24 seconds
Time spent in heuristics: 0.00 seconds
Status: 3000
Primal Bound: -107240
Dual Bound  : -107240
Gap (%)     : 0
Deleting DSP environment
```
So I believe the Gurobi is not solving the master problem correctly.
I was trying to solve this problem for several days. I located where the `iterations` is printed, it turns out it is in Alps library. From my understanding, all models and solvers' information are passed into Alps library for DW algorithm. More specifically, it is in the [rootsearch() function](https://github.com/kibaekkim/CHiPPS-ALPS/blob/d635ae3e2a044ce32eca8902cc0e38644a3db733/Alps/src/AlpsKnowledgeBrokerSerial.cpp#L191), and the rootsearch() function called the function [solveMaster ()](https://github.com/cheesecakeball/DSP/blob/fa2b5ec02ab53a2de5658c600478959ac7da04c0/src/Solver/DantzigWolfe/DwBundleDual.cpp#L415).
However, I still cannot figure out why I cannot get the optimal solution, why this error happens.

By the way,  in the information printed in each iteration, should it be `state 3000` instead of `statue 3000`?

kibaekkim commented on src/Solver/DantzigWolfe/DwBundleDual.cpp:
```diff
@@ -418,7 +420,8 @@ DSP_RTN_CODE DwBundleDual::solveMaster() {
 	/** call solver */
 	status_ = callMasterSolver();
-
+	printf("master solver suffccessully called");
```
please remove all these debugging lines.

kibaekkim commented on src/Solver/DantzigWolfe/DwSolverSerial.cpp:
```diff
@@ -95,7 +95,7 @@ DSP_RTN_CODE DwSolverSerial::solve() {
     alpsBroker.search(alps_);
 	//alpsBroker.printBestSolution();
-
+	printf("I am here 1\n");
```
Please remove all debugging lines.

kibaekkim commented on src/Solver/DantzigWolfe/DwSolverSerial.cpp:
```diff
@@ -128,11 +128,13 @@ DSP_RTN_CODE DwSolverSerial::solve() {
 	DspNodeSolution* solution = NULL;
 	if (alpsBroker.hasKnowledge(AlpsKnowledgeTypeSolution)) {
 		solution = dynamic_cast<DspNodeSolution*>(alpsBroker.getBestKnowledge(AlpsKnowledgeTypeSolution).first);
+		printf("I am here 3\n");
```
Please remove all debugging lines.

kibaekkim commented on src/Solver/DantzigWolfe/DwSolverSerial.cpp:
```diff
@@ -128,11 +128,13 @@ DSP_RTN_CODE DwSolverSerial::solve() {
 			if (model_->isStochastic() && bestprimsol_.size() > 0) {
 				TssModel* tss = dynamic_cast<TssModel*>(model_);
 				if (bestprimsol_.size() > 0)
+					printf("I am here 2\n");
```
Please remove all debugging lines.

kibaekkim commented on src/Utility/DspParams.cpp:
```diff
@@ -229,7 +222,25 @@ void DspParams::initIntParams()
 	IntParams_.createParam("DW/MASTER/SOLVER", OsiCpx);
 	IntParams_.createParam("DW/SUB/SOLVER", OsiCpx);
+#else
+	IntParams_.createParam("DW/MASTER/SOLVER", OsiClp);
```
Why is this set to `OsiClp`? The master problem is QP that can be solved by `OsiGrb`.

kibaekkim commented on test/CMakeLists.txt:
```diff
@@ -1,31 +1,45 @@
+add_test(NAME grb_noswot_dw COMMAND $ENV{PWD}/src/runDsp --algo dw --mps ${CMAKE_SOURCE_DIR}/examples/mps-dec/noswot.mps --dec ${CMAKE_SOURCE_DIR}/examples/mps-dec/noswot.dec --param ${CMAKE_SOURCE_DIR}/test/params_grb.txt --test -42)
+endif(GUROBILIB)
+
 if (MPI_CXX_FOUND)
```
Please also add mpi test with `grb`.

kibaekkim replied:
Please also add mpi test with `grb`.

cheesecakeball commented:
This is ready to review. 
All tests (with and without MPI) are passed. 
Fixed the behavior for DW.
Updated cmake files. 
Deleted garbage messages.

kibaekkim commented on src/Utility/DspParams.cpp:
```diff
@@ -223,14 +223,14 @@ void DspParams::initIntParams()
 	IntParams_.createParam("DW/SUB/SOLVER", OsiCpx);
 #else
-	IntParams_.createParam("DW/MASTER/SOLVER", OsiClp);
 #ifdef DSP_HAS_GRB
```
`BD/SUB/SOLVER` is missing for gurobi.

kibaekkim commented on extra/CMakeLists.txt:
```diff
@@ -73,7 +73,9 @@ else()
 
 if(GUROBILIB)
-	set(GUROBI_LIB "-L${GUROBI_LIB_DIR} -lgurobi${GUROBI_VERSION} -lm -lpthread")
+	get_filename_component(LIBGRB ${GUROBILIB} NAME_WE)
```
What do we get in `GUROBILIB` at this line? I am just curious.

cheesecakeball replied:
It is the full path of the Gurobi library. On my laptop is "/Library/gurobi901/mac64/lib/libgurobi90.dylib"

kibaekkim replied:
OK, then it should be enough to replace lines 76-78 with
```cmake
set(GUROBI_LIB "-l${GUROBILIB} -lm -lpthread")
```
Does gurobi also need to link `m` and `pthread`?

cheesecakeball replied:
I am pretty sure m is needed. But pthread is sometimes needed, depending on OS / compiler.

cheesecakeball replied:
> OK, then it should be enough to replace lines 76-78 with
> 
> ```cmake
> set(GUROBI_LIB "-l${GUROBILIB} -lm -lpthread")
> ```
> 
> Does gurobi also need to link `m` and `pthread`?
I also tried this. I got errors with this command
```
checking if user provides library for Gurobi... yes
configure: error: Cannot find file gurobi_c.h in /Users/yingqiuzhang
configure: error: /bin/sh '/Users/yingqiuzhang/Documents/DSPtest/DSP/extra/Osi/Osi/configure' failed for Osi
make[2]: *** [extra/Osi-prefix/src/Osi-stamp/Osi-configure] Error 1
make[1]: *** [extra/CMakeFiles/Osi.dir/all] Error 2
make: *** [all] Error 2

```

kibaekkim replied:
It seems related to this line.
```cmake
set(WITH_GUROBI_INCDIR "--with-gurobi-incdir=${GUROBI_INC_DIR}")
```

cheesecakeball replied:
> It seems related to this line.
> 
> ```cmake
> set(WITH_GUROBI_INCDIR "--with-gurobi-incdir=${GUROBI_INC_DIR}")
> ```


Sorry that I did not reproduce my error correctly, it is actually like this

```
configure: error: Cannot find symbol(s) GRBloadenv with GRB
configure: error: /bin/sh '/Users/yingqiuzhang/Documents/DSPtest/DSP/extra/Osi/Osi/configure' failed for Osi
make[2]: *** [extra/Osi-prefix/src/Osi-stamp/Osi-configure] Error 1
make[1]: *** [extra/CMakeFiles/Osi.dir/all] Error 2
make: *** [all] Error 2
```
